Add wrappers around public static getters & setters of QCoreApplication

ratijas opened the pull request:
To be honest, I'm a bit concerned about the name of this new module, because core is also a part of rustc std.

---

Closes #113

ogoffart commented on qmetaobject/src/core.rs:
```diff
@@ -0,0 +1,129 @@
+    /// Wrapper around [`void setApplicationName(const QString &application)`][qt] static method.
+    ///
+    /// [qt]: https://doc.qt.io/qt-5/qcoreapplication.html#applicationName-prop
+    pub fn set_application_name(application: &str) {
```
Why not take a QString since we take ownership as a QString?

ratijas replied:
AFAIK String to QString conversion is not in-place, and re-allocated memory. So, why drop a String if we only need a str slice? This follows minimal requirements rule.

Here are the conversion impls, btw
```rust
impl<'a> From<&'a str> for QString {
    /// Copy the data from a `&str`.
    fn from(s: &'a str) -> QString {
        let len = s.len();
        let ptr = s.as_ptr();
        cpp!(unsafe [len as "size_t", ptr as "char*"] -> QString as "QString" {
            return QString::fromUtf8(ptr, len);
        })
    }
}
impl From<String> for QString {
    fn from(s: String) -> QString {
        QString::from(&*s)
    }
}
```

ratijas replied:
Hilariously, documentation of [QString::fromUtf8](https://doc.qt.io/qt-5/qstring.html#fromUtf8) says nothing about ownership. It just "returns a QString initialized with the first size bytes of the UTF-8 string str" — whatever that "initialization" might imply. And the type of `const char *str` requiring it being const doesn't make the whole situation better. Here we go reading source code once again…

ratijas replied:
I think, the broad question here is: do we want to make wrappers more "Rust idiomatic" by exposing QString as String/&str etc., or do we want preserve the implicit sharing provided by Qt data types (not only strings)?

ogoffart replied:
Right, QString is owning the string.    So it allocates.  Just like `String` in rust.

> So, why drop a String if we only need a str slice?

We don't need a str slice,  we need a QString to give it to the application.

ratijas replied:
Shout-outs to #81, by the way. I think it was mentioned somewhere there, but was never resolved. If not, then it should be added.

So, give me a second to refactor it into QStrings.

ratijas commented:
I assume, this is good to merge then?

Free (like in beer) weekends are coming, and I hope to start hacking on new stuff, or rather splitting the old good macro into parser/generator steps.

ogoffart commented on qmetaobject/src/lib.rs:
```diff
@@ -195,6 +195,7 @@ pub use qttypes::*;
 pub use tablemodel::*;
 
 pub mod connections;
+pub mod core;
```
not sure core is a good name for a module because of the core crate.   Maybe core_application or just application would be better.

ratijas replied:
So one Rust module per "class" then? And what about Qt modules, i.e. QtCore?  I recall we never reached consensus on this topic either. I believe it is worth experimenting with both styles here and there, and see which one works best in the end.

How about _qtcore_ for now, and maybe refactor into _qtcore::core_application_ later if needed? At the very least, that would go in-line with _qtdeclarative_ and _qtquickcontrols2_ modules.

ogoffart replied:
yeah, i don't know if it makes sense to have one rust module per Qt library.  
maybe per group of class that belong together.

ratijas replied:
It is hard, and probably meaningless, to try to define a group of related classes. That would be some opinionated and not well-known mapping which would need to be documented somewhere.

On the other hand, file per class looks reasonable, because (1) that's what Qt already does, and (2) classes have so called "related non-members" and associated enums that would fit in such rust file/module just fine.

If there is ever going to be too many wrappers, at that point it might be feasible to split into "qmetaobject_qt{core,declarative,…}" crates to take advantage of parallel compilation in multi-threaded environments.

ogoffart reviewed (approved):
:+1: 
Yes, good to merge.
Althouhg maybe you want to rename the `core` module

ratijas commented:
Refactored into _qtcore::core_application_ module. Added a note to clarify that `quit()` method is not and should not have `!` never type, because it actually returns.

> Free (like in beer) weekends are coming, and I hope to start hacking on new stuff, or rather splitting the old good macro into parser/generator steps.

That… didn't age well. All I've really done is [fixed JSON.stringify](https://bugreports.qt.io/browse/QTBUG-95324) in Qt V4 JavaScript engine, learned few new tricks in Zelda Breath of the Wild — and there goes my weekend.
